overlord/snapshotstate/backend/backend.go

```diff
@@ -0,0 +1,256 @@
+// error isn't EOF, it's returned as the error of the iterator).
+func Iter(ctx context.Context, f func(*Reader) error) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	dir, err := dirOpen(dirs.SnapshotsDir)
+	if err != nil {
+		if osutil.IsDirNotExist(err) {
+			// no dir -> no snapshots
+			return nil
+		}
+		return fmt.Errorf("cannot open snapshots directory: %v", err)
+	}
+	defer dir.Close()
+
+	var names []string
+	for err == nil {
+		names, err = dirNames(dir, 100)
+		for _, name := range names {
+			if err = ctx.Err(); err != nil {
+				break
+			}
+
+			filename := filepath.Join(dirs.SnapshotsDir, name)
+			rsh, openError := open(filename)
+			// rsh can be non-nil even when openError is not nil (in
+			// which case rsh.Broken will have a reason). f can
+			// check and either ignore or return an error when
+			// finding a broken snapshot.
+			if rsh != nil {
+				err = f(rsh)
+			} else {
+				// TODO: use warnings instead
+				logger.Noticef("cannot open snapshot %q: %v", name, openError)
+			}
+			if openError == nil {
+				// if openError was nil the snapshot was opened and needs closing
+				if closeError := rsh.Close(); err == nil {
+					err = closeError
+				}
+			}
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	if err == io.EOF {
+		err = nil
+	}
+
+	return err
+}
+
+// List valid snapshots sets.
+func List(ctx context.Context, setID uint64, snapNames []string) ([]client.SnapshotSet, error) {
+	setshots := map[uint64][]*client.Snapshot{}
+	err := Iter(ctx, func(sh *Reader) error {
+		if setID == 0 || sh.SetID == setID {
+			if len(snapNames) == 0 || strutil.ListContains(snapNames, sh.Snap) {
+				setshots[sh.SetID] = append(setshots[sh.SetID], &sh.Snapshot)
+			}
+		}
+		return nil
+	})
+
+	sets := make([]client.SnapshotSet, 0, len(setshots))
+	for id, shots := range setshots {
+		sort.Sort(bySnap(shots))
+		sets = append(sets, client.SnapshotSet{ID: id, Snapshots: shots})
+	}
+
+	sort.Sort(byID(sets))
+
+	return sets, err
+}
+
+// Filename of the given client.Snapshot in this backend.
+func Filename(sh *client.Snapshot) string {
+	// this _needs_ the snap name and version to be valid
+	return filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_%s_%s.zip", sh.SetID, sh.Snap, sh.Version))
+}
+
+// Save a snapshot
+func Save(ctx context.Context, id uint64, si *snap.Info, cfg map[string]interface{}, usernames []string) (sh *client.Snapshot, e error) {
+	if err := os.MkdirAll(dirs.SnapshotsDir, 0700); err != nil {
+		return nil, err
+	}
+
+	sh = &client.Snapshot{
+		SetID:    id,
+		Snap:     si.Name(),
+		Revision: si.Revision,
+		Version:  si.Version,
+		Time:     time.Now(),
+		SHA3_384: make(map[string]string),
+		Size:     0,
+		Conf:     cfg,
+	}
+
+	aw, err := osutil.NewAtomicFile(Filename(sh), 0600, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return nil, err
+	}
+	// if things worked, we'll commit (and Cancel becomes a NOP)
+	defer aw.Cancel()
```

Comment:
atomic files have both Close and Cancel (Cancel closes, but not viceversa). I'm not sure we ever use Close on its own, and you're not the first person to ask this (and in fact we had a bug where an atomic file was closed instead of being committed or canceled), so it might be worth revisiting.